File: apps/ios/Milo Tales/Network/CharacterAPI.swift
//
//  CharacterAPI.swift
//  Milo Tales
//

import Foundation

nonisolated struct CreateCharacterRequest: Encodable {
    let role: CharacterRole
    let name: String
    let symbolName: String
    let tint: String
    let tagline: String?
    let age: Int?
    let gender: Gender?
    let hairColor: String?
    let eyeColor: String?
    let hairstyle: String?
    let interests: [String]
    let extraInterestNote: String
}

nonisolated struct UpdateCharacterRequest: Encodable {
    var role: CharacterRole? = nil
    var name: String? = nil
    var symbolName: String? = nil
    var tint: String? = nil
    var tagline: String? = nil
    var age: Int? = nil
    var gender: Gender? = nil
    var hairColor: String? = nil
    var eyeColor: String? = nil
    var hairstyle: String? = nil
    var interests: [String]? = nil
    var extraInterestNote: String? = nil
}

nonisolated struct CharacterResponse: Decodable {
    let id: String
    let role: CharacterRole
    let name: String
    let symbolName: String
    let tint: String
    let tagline: String?
    let age: Int?
    let gender: Gender?
    let hairColor: String?
    let eyeColor: String?
    let hairstyle: String?
    let interests: [String]
    let extraInterestNote: String
    let createdAt: String
    let updatedAt: String

    func toCharacter() -> Character {
        Character(
            id: UUID(uuidString: id) ?? UUID(),
            name: name,
            role: role,
            symbolName: symbolName,
            tintName: tint,
            tagline: tagline ?? "",
            age: age,
            gender: gender,
            hairColor: hairColor,
            eyeColor: eyeColor,
            hairstyle: hairstyle,
            interests: interests,
            extraInterestNote: extraInterestNote
        )
    }
}

enum CharacterAPI {
    static func list() async throws -> [Character] {
        let payloads: [CharacterResponse] = try await APIClient.shared.get("/api/v1/characters")
        return payloads.map { $0.toCharacter() }
    }

    static func create(_ input: CreateCharacterRequest) async throws -> Character {
        let payload: CharacterResponse = try await APIClient.shared.post(
            "/api/v1/characters",
            body: input
        )
        return payload.toCharacter()
    }

    static func update(_ id: UUID, _ patch: UpdateCharacterRequest) async throws -> Character {
        let payload: CharacterResponse = try await APIClient.shared.patch(
            "/api/v1/characters/\(id.uuidString.lowercased())",
            body: patch
        )
        return payload.toCharacter()
    }

    static func delete(_ id: UUID) async throws {
        struct DeletedResponse: Decodable { let id: String }
        let _: DeletedResponse = try await APIClient.shared.delete(
            "/api/v1/characters/\(id.uuidString.lowercased())"
        )
    }
}

File: apps/android/app/src/main/java/com/cortexlumora/lunastories/network/CharacterAPI.kt
package com.cortexlumora.lunastories.network

import io.ktor.client.request.setBody
import io.ktor.http.ContentType
import io.ktor.http.HttpMethod
import io.ktor.http.contentType
import kotlinx.serialization.Serializable

enum class CharacterRole { main, side }
enum class Gender { male, female, na }
enum class CharacterRelation {
    parent, grandparent, friend, pet, sibling, teacher, imaginary, other;

    val displayName: String get() = when (this) {
        parent -> "Parent"
        grandparent -> "Grandparent"
        friend -> "Friend"
        pet -> "Pet"
        sibling -> "Sibling"
        teacher -> "Teacher"
        imaginary -> "Imaginary"
        other -> "Other"
    }
}

@Serializable
data class CharacterResponse(
    val id: String,
    val role: CharacterRole,
    val name: String,
    val symbolName: String,
    val tint: String,
    val tagline: String? = null,
    val relation: CharacterRelation? = null,
    val customRelation: String? = null,
    val age: Int? = null,
    val gender: Gender? = null,
    val hairColor: String? = null,
    val eyeColor: String? = null,
    val hairstyle: String? = null,
    val interests: List<String> = emptyList(),
    val extraInterestNote: String = "",
    val createdAt: String,
    val updatedAt: String,
)

@Serializable
data class CreateCharacterRequest(
    val role: CharacterRole,
    val name: String,
    val symbolName: String,
    val tint: String,
    val tagline: String? = null,
    val relation: CharacterRelation? = null,
    val customRelation: String? = null,
    val age: Int? = null,
    val gender: Gender? = null,
    val hairColor: String? = null,
    val eyeColor: String? = null,
    val hairstyle: String? = null,
    val interests: List<String> = emptyList(),
    val extraInterestNote: String = "",
)

@Serializable
data class UpdateCharacterRequest(
    val name: String? = null,
    val symbolName: String? = null,
    val tint: String? = null,
    val tagline: String? = null,
    val relation: CharacterRelation? = null,
    val customRelation: String? = null,
    val age: Int? = null,
    val gender: Gender? = null,
    val hairColor: String? = null,
    val eyeColor: String? = null,
    val hairstyle: String? = null,
    val interests: List<String>? = null,
    val extraInterestNote: String? = null,
)

object CharacterAPI {
    suspend fun list(): List<CharacterResponse> = APIClient.get("/characters")

    suspend fun create(input: CreateCharacterRequest): CharacterResponse =
        APIClient.request("/characters") {
            method = HttpMethod.Post
            contentType(ContentType.Application.Json)
            setBody(input)
        }

    suspend fun update(id: String, patch: UpdateCharacterRequest): CharacterResponse =
        APIClient.request("/characters/$id") {
            method = HttpMethod.Patch
            contentType(ContentType.Application.Json)
            setBody(patch)
        }

    suspend fun delete(id: String) {
        APIClient.request<DeleteResponse>("/characters/$id") {
            method = HttpMethod.Delete
        }
    }

    @Serializable
    private data class DeleteResponse(val id: String)
}
